packages/gen-ai/bff/internal/integrations/kubernetes/token_k8s_client.go

```diff
@@ -1275,3 +1275,193 @@ func (kc *TokenKubernetesClient) DeleteLlamaStackDistribution(ctx context.Contex
 	kc.Logger.Info("successfully deleted LlamaStackDistribution", "namespace", namespace, "name", targetLSD.Name, "displayName", name)
 	return targetLSD, nil
 }
+
+// GetModelProviderInfo retrieves provider configuration for a model from LSD configmap
+func (kc *TokenKubernetesClient) GetModelProviderInfo(ctx context.Context, identity *integrations.RequestIdentity, namespace string, modelID string) (*ModelProviderInfo, error) {
+	// Get LlamaStackDistribution
+	lsdList, err := kc.GetLlamaStackDistributions(ctx, identity, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get LlamaStackDistributions: %w", err)
+	}
+
+	if len(lsdList.Items) == 0 {
+		return nil, fmt.Errorf("no LlamaStackDistribution found in namespace %s", namespace)
+	}
+
+	lsd := lsdList.Items[0]
+
+	// Get configmap name
+	configMapName := "llama-stack-config"
+	if lsd.Spec.Server.UserConfig != nil && lsd.Spec.Server.UserConfig.ConfigMapName != "" {
+		configMapName = lsd.Spec.Server.UserConfig.ConfigMapName
+	}
+
+	// Retrieve configmap
+	configMap, err := kc.GetConfigMap(ctx, identity, namespace, configMapName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get configmap: %w", err)
+	}
+
+	runYAML, ok := configMap.Data["run.yaml"]
+	if !ok {
+		return nil, fmt.Errorf("run.yaml not found in configmap")
+	}
+
+	// Parse model and provider info from YAML
+	return kc.parseModelProviderFromYAML(runYAML, modelID)
+}
+
+// parseModelProviderFromYAML parses the LlamaStack configuration YAML to extract model provider configuration
+// This is a two-step process:
+// 1. Find the model in the models section and get its provider_id
+// 2. Find that provider_id in the providers.inference section and get provider_type and url
+func (kc *TokenKubernetesClient) parseModelProviderFromYAML(runYAML string, modelID string) (*ModelProviderInfo, error) {
+	lines := strings.Split(runYAML, "\n")
+
```

Comment:
Can we use a map or 2 maps to avoid looping twice ?

Reply:
maybe we will not have better efficiency; but I think we should load it into yaml or json; and then extract the data using jsonpath or yaml equivalent.
That would be more maintainable code.